src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -97,8 +123,24 @@ public CachingContext(JsonSerializerOptions options)
             // Property only accessed by reflection in testing -- do not remove.
             // If changing please ensure that src/ILLink.Descriptors.LibraryBuild.xml is up-to-date.
             public int Count => _converterCache.Count + _jsonTypeInfoCache.Count;
-            public JsonConverter GetOrAddConverter(Type type) => _converterCache.GetOrAdd(type, Options.GetConverterFromType);
-            public JsonTypeInfo GetOrAddJsonTypeInfo(Type type) => _jsonTypeInfoCache.GetOrAdd(type, Options.GetJsonTypeInfoFromContextOrCreate);
+            public JsonConverter GetOrAddConverter(Type type) => _converterCache.GetOrAdd(type, Options.GetConverterFromTypeInfoOrOptionsNotCached);
+
+            public JsonTypeInfo? GetOrAddJsonTypeInfo(Type type)
+            {
+                if (_jsonTypeInfoCache.TryGetValue(type, out JsonTypeInfo? typeInfo))
```

Comment:
Why is this double lookup necessary? Can't `Options.GetJsonTypeInfoFromResolver` be embedded in the factory inside the `GetOrAdd` delegate?

Reply:
Options.GetJsonTypeInfoFromResolver returns nullable type info and I do need nullable type info in the GetConverter method. Also we don't want to cache null because null means we should retry since resolver for context can mutate (due to fallback)